File: ruby/beer-song/beer_song.rb
class BeerSong

  def verse(num)
    prefix(num) + suffix(num)
  end

  def verses(start, stop)
    result = ""
    start.downto(stop).each do |verse_num|
      result += verse(verse_num) + "\n"
    end
    result
  end

  def sing
    verses(99, 0)
  end

  def prefix(num)
    "#{num_or_No_more(num)} #{bottle_or_bottles(num)} of beer on the wall, "\
    "#{num_or_no_more(num)} #{bottle_or_bottles(num)} of beer.\n"
  end

  def suffix(num)
    if num == 0
      "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    else
      "Take #{one_or_it(num)} down and pass it around, #{num_or_no_more(num-1)} "\
      "#{bottle_or_bottles(num-1)} of beer on the wall.\n"
    end
  end

  def bottle_or_bottles(num)
    num == 1 ? "bottle" : "bottles"
  end

  def one_or_it(num)
    num != 1 ? "one" : "it"
  end

  def num_or_No_more(num)
    num != 0 ? num : "No more"
  end

  def num_or_no_more(num)
    num != 0 ? num : "no more"
  end
end

File: javascript/beer-song/beer-song.js
var Beer = (function() {
  "use strict";

  function verse(num) {
    if (num === 0) {
      return (
        "No more bottles of beer on the wall, no more bottles of beer.\n" +
        "Go to the store and buy some more, " +
        "99 bottles of beer on the wall.\n"
      );
    } else if (num === 1) {
      return (
        "1 bottle of beer on the wall, 1 bottle of beer.\n" +
        "Take it down and pass it around, " +
        "no more bottles of beer on the wall.\n"
      );
    } else {
      return (
        num +
        " bottles of beer on the wall, " +
        num +
        " bottles of beer.\nTake one down and pass it around, " +
        (num - 1) +
        " bottles of beer on the wall.\n"
      );
    }
  }

  function sing(from, to) {
    to = to || 0;
    var verses = [];

    for (var i = from; i >= to; i--) {
      verses.push(this.verse(i));
    }

    return verses.join("\n");
  }

  return {
    verse: verse,
    sing: sing
  };
})();

module.exports = Beer;
